src/deploy/kubernetes_functions.js

```diff
@@ -1,260 +1,292 @@
+            return this.kubectl(`config view --minify --output 'jsonpath={..namespace}'`, { ignore_namespace: true });
+        }
     }
 
-    async create_agent(params) {
-        const {
-            namespace,
-            // num_agents,
-            yaml_path = "./agent.yaml"
-        } = params;
-        const run_with_namespace = await this.run_with_namespace(namespace);
-        console.log(`Creating agent in kubernetes`);
-        await this.get_agent_yaml_from_server({ namespace, yaml_path });
-        //TODO: edit the yaml if needed 
-        //TODO: Fix the yaml for the correct image if needed.
-        //await this.edit_number_of_agents(...);
-        console.log(`Applying the agent yaml ${yaml_path}`);
-        await promise_utils.exec(`kubectl apply -f ${yaml_path} ${run_with_namespace}`);
-    }
+    update_statefulset({ statefulset, replicas, image, envs, cpu, mem, pv }) {
+        if (image) {
+            // modify image of the statefulset
+            statefulset.spec.template.spec.containers[0].image = image;
+        }
 
-    async edit_image_in_yaml(params) {
-        const {
-            yaml,
-            // num_agents
-        } = params;
-        const file_content = fs.readFileSync(yaml);
-        console.log(file_content);
-    }
+        if (cpu) {
+            statefulset.spec.template.spec.containers[0].resources.requests.cpu = cpu;
+        }
 
-    async edit_number_of_agents(params) {
-        const {
-            agent_yaml,
-            // num_agents
-        } = params;
-        await this.editYaml(agent_yaml);
+        if (mem) {
+            statefulset.spec.template.spec.containers[0].resources.requests.memory = mem;
+        }
+
+        // set env 
+        if (envs) {
+            statefulset.spec.template.spec.containers[0].env = (statefulset.spec.template.spec.containers[0].env || []).concat(envs);
+        }
+
+        if (replicas) {
+            statefulset.spec.replicas = replicas;
+        }
+
+        if (!pv) {
+            //remove persistent volume claim and mounts from the statefulset
+            statefulset.spec.template.spec.containers[0].volumeMounts = null;
+            statefulset.spec.volumeClaimTemplates = null;
+
+        }
     }
 
-    async delete_pvc(params) {
-        const {
-            namespace
-        } = params;
-        if ((!namespace)) {
-            throw new Error('In order to protect other namespaces, namespace param is mandatory');
+
+    async deploy_server({ image, server_yaml, envs, cpu, mem, pv }) {
+        const server_details = {};
+        try {
+            let resources_file_path = path.join(this.output_dir, `${this._namespace}.server_deployment.${Date.now()}.json`);
+            // modify resources and write to temp yaml
+            const resources = await this.read_resources(server_yaml);
+            const statefulset = resources.find(res => res.kind === 'StatefulSet');
+            const pod_name = `${statefulset.metadata.name}-0`;
```

Comment:
We can write a function that gets the statfulset name and returns it as we use this code i 2 places.
```
            const resources = await this.read_resources(server_yaml);
            const statefulset = resources.find(res => res.kind === 'StatefulSet');
            const pod_name = `${statefulset.metadata.name}-0`;
```

Reply:
I don't think there is a lot of value in creating a function for that. we don't save a lot and there are small differences anyway that we will need to handle. in my opinion it isn't worth it for just these two places